src/phillips_dessouky.rs

```diff
@@ -0,0 +1,84 @@
+use pyo3::prelude::*;
+
+use ordered_float::OrderedFloat;
+use pathfinding::directed::edmonds_karp::{
+    SparseCapacity,
+    Edge,
+    edmonds_karp
+};
+
+use std::time::{
+    Instant,
+    Duration,
+};
+use log::info;
+
+
+#[pyclass]
+pub struct PhillipsDessouky {
+    #[pyo3(get)]
+    node_ids: Vec<u32>,
+    source_node_id: u32,
+    sink_node_id: u32,
+    edges_raw: Vec<((u32, u32), f64)>,
+}
+
+#[pymethods]
+impl PhillipsDessouky {
+    #[new]
+    fn new(
+        node_ids: Vec<u32>,
+        source_node_id: u32,
+        sink_node_id: u32,
+        edges_raw: Vec<((u32, u32), f64)>,
+    ) -> PyResult<Self> {
+        Ok(PhillipsDessouky {
+            node_ids,
+            source_node_id,
+            sink_node_id,
+            edges_raw,
+        })
+    }
+
+    fn max_flow(&self) -> PyResult<Vec<((u32, u32), f64)>> {
+        let profiling_start = Instant::now();
+        let edges_edmonds_karp: Vec<Edge<u32, OrderedFloat<f64>>> = self.edges_raw.iter().map(|((from, to), cap)| {
+            ((*from, *to), OrderedFloat(*cap))
+        }).collect();
+        let profiling_end = Instant::now();
+        info!("PROFILING Rust_PhillipsDessouky::max_flow scaling to OrderedFloat time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
+
+        let profiling_start = Instant::now();
+        let (flows, _max_flow, _min_cut) = edmonds_karp::<_, _, _, SparseCapacity<_>>(
+            &self.node_ids,
+            &self.source_node_id,
+            &self.sink_node_id,
+            edges_edmonds_karp,
+        );
+        let profiling_end = Instant::now();
+        info!("PROFILING Rust_PhillipsDessouky::max_flow edmonds_karp time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
+
+        let profiling_start = Instant::now();
+        let flows_f64: Vec<((u32, u32), f64)> = flows.iter().map(|((from, to), flow)| {
+            ((*from, *to), flow.into_inner())
+        }).collect();
+        let profiling_end = Instant::now();
+        info!("PROFILING Rust_PhillipsDessouky::max_flow reformat OrderedFloat to f64 time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
```

Comment:
These `PROFILING` log messages are for dev purposes, and they need to be removed/prettified when we land it on `master`.

Reply:
Agreed. I will keep them until the end of this chain, then remove them. I think having a profiling integrated into lowtime somehow could be a cool, separate project. I've made a note of this in the "Design Decisions" section of the main PR comment/overview as well.

Reply:
Maybe try running something like [Samply](https://github.com/mstange/samply/) on the program. For debug symbols I think you'll have to put `debug = true` under `[profile.release]` to get a flamegraph.

Reply:
I haven't tried using Samply yet, but just based on running with/without Rust logger the profiling results look pretty much the same. In fact, for some runs, it seems to run slower without the Rust logger(?) I think this may just be inherent randomness involved in profiling large programs. and that we can assume there is no significant overhead.